### PR TITLE
[GEOS-10651] Incorporate Vector Mosaic Datastore

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-10651](https://osgeo-org.atlassian.net/browse/GEOS-10651)

docs

cleanup

more cleanup

more cleanup again

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Review by aaime — CHANGES_REQUESTED** [review]
Testing the store I've also found that "compute bounds" is not working and returns nothing (should just accumulate the bounds from the index table?).

In addition, I've created a index shapefile with a "name" column that I could use for selection, which more or less contains the file name:

```ogr2ogr places.shp places_raw.shp -dialect SQLite -sql "SELECT Geometry,'file://'||SUBSTR(params,1,LENGTH(params)-2) AS params, replace(params, rtrim(params, replace(params, '/', '')), '') as name from places_raw"```

Now, when I try to use it, things are not working as expected:
* If I specify a filter with no match, I get an error "no delegate features found". Should not be an error, I should just get nothing back, opening no delegate files in the process
* If I specify a filter that actually matches some index records, I get no error, but also no features back, just an empty map.

To do the tests I've used the PLACES shapefiles from Tiger 2021:
* The filter ``name like '%03%'`` returns the error.
* The filter ``name like '%2021_48%'`` should match features from one state, but gives me back nothing

The stack trace from the error case:

```
Caused by: java.lang.IllegalArgumentException: No delegate features found
	at org.geotools.vectormosaic.VectorMosaicFeatureReader.getGranuleType(VectorMosaicFeatureReader.java:82)
	at org.geotools.vectormosaic.VectorMosaicFeatureReader.<init>(VectorMosaicFeatureReader.java:66)
	at org.geotools.vectormosaic.VectorMosaicFeatureSource.getReaderInternal(VectorMosaicFeatureSource.java:176)
	at org.geotools.data.store.ContentFeatureSource.getReader(ContentFeatureSource.java:636)
	at org.geotools.data.store.ContentFeatureCollection.features(ContentFeatureCollection.java:173)
	at org.geotools.data.store.ContentFeatureCollection.features(ContentFeatureCollection.java:52)
	at org.geotools.data.crs.ForceCoordinateSystemFeatureResults.openIterator(ForceCoordinateSystemFeatureResults.java:93)
	at org.geotools.feature.collection.AbstractFeatureCollection.features(AbstractFeatureCollection.java:66)
	at org.geotools.feature.collection.AbstractFeatureCollection.features(AbstractFeatureCollection.java:50)
	at org.geotools.renderer.lite.StreamingRenderer.drawPlain(StreamingRenderer.java:2651)
	at org.geotools.renderer.lite.StreamingRenderer.processStylers(StreamingRenderer.java:2286)
	at org.geotools.renderer.lite.StreamingRenderer.paint(StreamingRenderer.java:919)
	at org.geoserver.wms.map.RenderedImageMapOutputFormat.produceMap(RenderedImageMapOutputFormat.java:545)
	... 133 more
```

File: doc/en/user/source/community/vector-mosaic/installing.rst [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+Installing Vector Mosaic Datastore
+==================================
+
+To install the Vector Mosaic datastore:
+
+#. Download the module: :download_community:`vector-mosaic`
+
+#. Extract this file and place the JARs in ``WEB-INF/lib``. 
+
+#. Perform any configuration required by your servlet container, and then restart. On startup, JDBCStore will create a configuration directory ``vector-mosaic`` in the :ref:`datadir` .
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/user/source/community/vector-mosaic/installing.rst]: fixed

File: doc/en/user/source/community/vector-mosaic/installing.rst~ [hunk 2]
```diff
@@ -0,0 +1,14 @@
+.. _community_vector_mosaic_installing:
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/user/source/community/vector-mosaic/installing.rst~]: fixed

File: doc/en/user/source/community/vector-mosaic/delegate.rst [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+Any delegate parameters beyond the two required can serve as queryable/filterable attributes used to narrow the number of potential granule vectors that are searched by a query.  The non-required parameters will be combined with the vector granule parameters to create the output featuretype.
+
+The `ogrtindex <https://gdal.org/programs/ogrtindex.html>`_ commandline tool from the GDAL library can be used to create a simple delegate layer.  Here is an example that generates a delegate shapefile from a directory of shapefiles.  Note that the third step below is to use ogr2ogr commandline to trim a comma and number that ogrtindex appends to the end of the granule reference, rename the reference column to "params" and to append the URI identifier to the front of the reference.
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/user/source/community/vector-mosaic/delegate.rst]: fixed

File: doc/en/user/source/community/vector-mosaic/delegate.rst [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+#. Switch to directory with the shapefiles
+#. ogrtindex  -write_absolute_path delegate_raw.shp \*.shp
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/user/source/community/vector-mosaic/delegate.rst]: fixed

**Comment by turingtestfail** [comment]
> Caused by: java.lang.IllegalArgumentException: No delegate features found

Fix for this just pushed to the GeoTools PR.  I was using the filtered delegate featurecollection to grab a sample granule to generate a featuretype when I should have been using an unfiltered collection.

The compute bounds is returning nothing due to a performance optimization that returns null for the getBoundsInternal call if the index part of the query equals Filter.INCLUDE.  I can change this to instead return the index bounds if that is preferred.

**Comment by aaime** [comment]
getBoundsInternal should just accumulate the bounds of the index, without opening the secondary files... as long as the query hits only the index fields, you can run it efficiently.

File: doc/en/user/source/community/vector-mosaic/configuration.rst [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+   * - ``connectionParameterKey``
+     - The delegate store has a mandatory field called "params". Params can either be a URI pointing at the granule resource location or it can be a configuration string in .properties format. (See `Java Properties file <https://en.wikipedia.org/wiki/.properties>`_ for more details about the format) In the latter case this optional parameter specifies which key points at the location of the granule.  Accepted values are "file" and "url".
+   * - ``preferredDataStoreSPI``
+     - This optional parameter can serve as an optimization to speed up the lookup of granule data stores.  Instead of attempting to use the mandatory delegate params field (See `delegate requirements <delegate.html>`_ for more details about delegate store requirements.) to look up supported data store types, the Vector Mosaic data store will use the data store SPI specified in this field to identify the correct type.  The sample value is "org.geotools.data.shapefile.ShapefileDataStoreFactory".  Other possible values can be found in the docs for `DataStoreFactorySpi <https://docs.geotools.org/latest/javadocs/org/geotools/data/DataStoreFactorySpi.html>`_
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/user/source/community/vector-mosaic/configuration.rst]: Done.

**Comment by turingtestfail** [comment]
> getBoundsInternal should just accumulate the bounds of the index, without opening the secondary files... as long as the query hits only the index fields, you can run it efficiently.

Done and pushed to GeoTools PR.

File: doc/en/user/source/community/vector-mosaic/delegate.rst [hunk 6]
```diff
@@ -0,0 +1,18 @@
+.. _community_vector_mosaic_delegate:
+
+Vector Mosaic Datastore Delegate Requirements
+=============================================
+
+The Vector Mosaic Datastore Delegate is a datastore that contains references to the vector granule datastores, bounding polygon or multipolygon geometry to delineate the index area, and optionally other attributes that can be queried in order to return vector granules.
+
+The delegate datastore can be in any format that GeoServer supports but there are two required fields.  There must be a geometry field representing the index spatial area in either Polygon or MultiPolygon form.  There also must be a field called "params" in text format that contains either URIs pointing at granule resources like shapefiles -or- a configuration string in .properties format. (See `Java Properties file <https://en.wikipedia.org/wiki/.properties>`_ for more details about the format).
```
[the remark this answers is not in the document]
Comment (aaime) [reply doc/en/user/source/community/vector-mosaic/delegate.rst:8]: This change seems to have been ignored...

File: doc/en/user/source/community/vector-mosaic/delegate.rst [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+Any delegate parameters beyond the two required can serve as queryable/filterable attributes used to narrow the number of potential granule vectors that are searched by a query.  The non-required parameters will be combined with the vector granule parameters to create the output featuretype.
```
[the remark this answers is not in the document]
Comment (aaime) [reply doc/en/user/source/community/vector-mosaic/delegate.rst:10]: So this this one... parameters are the options you give to a store. "Field" or "Attribute" are the columns of the data itself.

File: doc/en/user/source/community/vector-mosaic/configuration.rst [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+=====================================
+
+When the extension has been installed, `Vector Mosaic Data Store` will be an option in the `Vector Data Sources` list when creating a new data store.
+
+.. figure:: images/vector-mosaic-vector-create.png
+   :align: center
+
+   *Vector Mosaic Data Store in the list of vector data sources*
+
+.. figure:: images/vector-mosaic-vector-configure.png
```
Comment (aaime) [inline doc/en/user/source/community/vector-mosaic/configuration.rst:13]: Outdated screenshot? I don't see drop downs.

**Review by aaime — CHANGES_REQUESTED** [review]

File: doc/en/user/source/community/vector-mosaic/configuration.rst (hunk 8 above)
Comment (aaime) [reply doc/en/user/source/community/vector-mosaic/configuration.rst:13]: Also, looking at the current panel:

image

- The second drop-down is not auto-fill type
- The contents inside the second drop-down should be alphabetically sorted, helps finding stuff in it when there are many stores.

Comment (turingtestfail) [reply doc/en/user/source/community/vector-mosaic/configuration.rst:13]: Fixed

**Pull request merged**